File: python/scripts/game.js
class Game{

    constructor(players){

        this.dominoes = []
        this.players = players

        for(let i=0; i<7; i++){

            for(let j = i; j<7; j++)

                this.dominoes.push(new Domino(i,j))

        }

        this.distribute()

        this.players = this.start()

    }

    distribute(){


        var copy = this.dominoes

        for (let i=0; i<7; i++){

            for(let player of this.players){

                var index = Math.floor(Math.random() * copy.length)
                var bone = copy[index]
                player.hand.add(bone)
                bone.site = player

                copy.splice(index,1)

            }
        } 
        for(let player of this.players){

            console.log("Peças de " +player.nome+":")
            player.hand.display()

        }

        this.pile = new Pile(copy)

    }

    start(){

        for(let i=6; i>=0 ;i--){

            for (let player of this.players){

                var domino = player.hand.get(i+":"+i)

                if(domino){

                    console.log(player.nome+" inicia com o carrão de "+i)
                    this.table = new Table(domino)
                    return player.next

                }
            }
        }
    }
    play(domino, place){

        if (this.player.hand.check(domino)){

            var domino = this.player.hand.get(domino)

            if (place && this.table.add_tail(domino) || ! place && this.table.add_head(domino)){
            
                if (this.player.hand.isEmpty()) return 2

                this.player = this.player.next
                return 1
        
            }

            else{

                this.player.hand.add(domino)
                return 0

            }

        }
        else return -1

    }
    buy(){

        var domino = this.pile.take()

        if (domino){

            this.player.hand.add(domino)

        }
        return 0
    }
}   

File: python/game/game.py
from nodes.domino import Domino
from structures.pile import Pile
from structures.table import  Table

from random import choice


class Game:

    
    #Estrutura básica do jogo de Dominó
    

    def __init__(self, players : list):

        self.dominoes = []
        self.players = players


        for i in range(7):

            for j in range(i,7):

                self.dominoes.append(Domino(i,j))
        
        self.distribute()

        self.player = self.start()

        print("\nAgora é a vez de",self.player.nome)



    def distribute(self):

        #Distribui os dominós entre os players
        

        copy = self.dominoes[:]

        for i in range(7):

            for player in self.players:

                bone = choice(copy)
                player.hand.add(bone)
                bone.site = player

                copy.remove(bone)

        for player in self.players:

            print("\nPeças de {} :\n".format(player.nome))
            player.hand.display()


        self.pile = Pile(copy)


    def start(self):

        #Seleciona o player com o maior 'carrão' para começar
        

        for i in range(0,7)[::-1]:


            for player in self.players:

                domino = player.hand.get("{}:{}".format(i,i))

                if domino:

                    print("\n{} inicia com carrão de {} !".format(player.nome,i))
                    self.table = Table(domino)
                    return player.next

    def play(self, domino : str, place : bool):

        #Realiza (ou tenta realizar) uma jogada
        

        if self.player.hand.check(domino):

            domino = self.player.hand.get(domino)

            if place and self.table.add_tail(domino) or not place and self.table.add_head(domino):

                if self.player.hand.isEmpty(): return 2

                self.player = self.player.next
                return 1


            else:

                self.player.hand.add(domino)
                return 0

        else: return  -1

    def buy(self):

        #Permite que o jogador compre um dominó da pilha, caso haja

        domino = self.pile.take()

        if domino:

            self.player.hand.add(domino)
            return domino

        return 0

